[FC-0036] feat: Add ContentTagsDrawer widget

yusuf-musleh opened the pull request:
## Description

This PR implements a side drawer widget for units' taxonomy tags. It includes displaying the Unit's tags, along with their lineage (ancestor tags) data. It also implements the listing the available taxonomy tags (including nesting ones) to select from to apply to this unit.

### Things to Note:
- The actual implementation for updating/edit the tags that belong to a unit will be implemented in the future PR.
- The way the nested taxonomy tags are fetched currently is a bit aggressive, makes a lot of requests, prefetching the data before it is needed/displayed. In interested of getting this PR out, I opted to update the implementation in a later PR, however if it a blocking issue in the review, I'll address it as part of this PR.

### Screenshots

![Screen Shot 2023-11-08 at 4 09 09 PM](https://github.com/openedx/frontend-app-course-authoring/assets/6829768/d4fa397c-bde6-4f59-a181-b4ed45126711)

![Screen Shot 2023-11-08 at 4 09 49 PM](https://github.com/openedx/frontend-app-course-authoring/assets/6829768/10f648ea-6741-4a7f-a46f-425674fe9d01)

![Screen Shot 2023-11-08 at 4 10 34 PM](https://github.com/openedx/frontend-app-course-authoring/assets/6829768/e1c68b2e-6865-410d-a261-412f782d267e)

## Supporting information

Related Tickets:
- https://github.com/openedx/modular-learning/issues/118

Related PRs:
- https://github.com/openedx/edx-platform/pull/33563

## Testing instructions

1. Make sure you followed the instructions mentioned in https://github.com/openedx/edx-platform/pull/33563 first
2. Check that the unit name correctly shows at the top of the drawer
3. Check that all the taxonomies are shown in the drawer
4. Check that a number badge is show next to taxonomies that have tags (the count represents the explicit tags only), if there are not tags applied, no badge should appear
5. Check that when you expand a section, it shows the tags that are applied to the unit, along with their ancestors if any (the x button currently does nothing, also part of a future PR)
6. Check that when clicking on "Add tags" button it a dropdown shows with the tags along with arrows if they have further nested tags. Clicking on the arrow, shows/hides the subtags. 
7. Check that you can click on the checkboxes (as of now, they do not do anything, just placeholders for a future PR)
8. Check that you can navigate the drawer using only the keyboard

---
Private-ref: [FAL-3523](https://tasks.opencraft.com/browse/FAL-3523)

openedx-webhooks commented:
Thanks for the pull request, @yusuf-musleh!  Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket as you can:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).


This is currently a draft pull request. When it is ready for our review and all tests are green, click "Ready for Review", or remove "WIP" from the title, as appropriate.

<!-- comment:end_of_wip -->




<!-- comment:external_pr -->
<!-- data: eyJkcmFmdCI6IHRydWV9 -->

mphilbrick211 commented:
Hi @yusuf-musleh! Would you be able to provide detail / additional context for this contribution? You can put this information at the top of the PR. Thanks!

yusuf-musleh commented:
@mphilbrick211 Sure thing! I've updated the PR with some information for context. I will add the full PR description once its closer to review, as it is still a work in progress.

e0d commented:
@yusuf-musleh looks like there are some coverage issues, can you have a look?

bradenmacdonald commented:
@mphilbrick211 This is actually part of FC-0036. Do you know why there's no "FC" label for PRs in this repo?

mphilbrick211 commented:
> @mphilbrick211 This is actually part of FC-0036. Do you know why there's no "FC" label for PRs in this repo?

I'm not sure. I don't have the ability to add the labels across repos, so I'll need to check on that internally.

bradenmacdonald commented on src/unit-taxonomy-tags-drawer/UnitTaxonomyTagsDrawer.jsx:
```diff
@@ -0,0 +1,102 @@
+  };
+
+  const useTaxonomyListData = () => {
+    const taxonomyListData = useTaxonomyListDataResponse(org);
```
I don't think the "taxonomy list" is actually necessary for the read-only view, though it's fine that you've implemented it because we'll need it in the editable version.

For the read-only version, the "get object tags" API returns the `name` (Taxonomy name) for each tag, which should be enough to display the view. (For the read-only version, we don't need to display taxonomies that don't yet have any tags.) And for displaying the ancestor tags, I'm working on an update to the API to make that possible, so it can be out of scope for now.

bradenmacdonald replied:
In fact, I think we don't need to wait for the taxonomy list to load before displaying anything here. We could just call "get object tags" and immediately render whatever tags+taxonomies it returns, and at the same time (or later) we can call the "list taxonomies" and then for any taxonomies which don't yet have tags, we could append them to the list. No need to change this now though.

yusuf-musleh replied:
Got it, thanks for the pointers, I'll keep this in mind when implementing the editing editable version.

bradenmacdonald commented on src/index.jsx:
```diff
@@ -53,10 +54,16 @@ const App = () => {
+              />
+              <Route
+                path="/tagging/components/widget/:unitId"
+                element={<UnitTaxonomyTagsDrawer />}
```
In the (very near) future, this will be used for components too (not just units), so maybe `ContentTagsDrawer` would be a better name.

bradenmacdonald commented on src/unit-taxonomy-tags-drawer/TagBubble.test.jsx:
```diff
@@ -0,0 +1,86 @@
+};
+
+const TagBubbleComponent = ({ value, subTagsCount, implicit }) => (
+  <AppProvider store={store}>
```
Since we are using React Query instead of redux for this stuff, you actually don't need `<AppProvider>` here in your test suite at all :)

bradenmacdonald commented on src/unit-taxonomy-tags-drawer/TagBubble.test.jsx:
```diff
@@ -0,0 +1,86 @@
+      },
+    });
+    store = initializeStore();
+  });
```
You can remove this whole `beforeEach` function - this code is only needed if you're using redux state, because redux uses a big giant global state, which is a pain for testing. With react-query, state is more isolated and you can skip all this mocking and boilerplate. I confirmed that this test suite passes without it.

bradenmacdonald commented on src/unit-taxonomy-tags-drawer/TaxonomyTagsCollapsible.jsx:
```diff
@@ -0,0 +1,116 @@
+
+import TagBubble from './TagBubble';
+
+const TaxonomyTagsCollapsible = ({ taxonomyAndTagsData }) => {
```
Please add a JSDoc-style comment above this component that gives a (short) description of what it is and what it does. (And what `taxonomyAndTagsData` is)

bradenmacdonald commented on src/unit-taxonomy-tags-drawer/TaxonomyTagsCollapsible.jsx:
```diff
@@ -0,0 +1,116 @@
+              let level = -1; // Starting from -1 since we increment at beginning of loop
+              return unitTag.lineage.map((lineageTag) => {
+                level += 1;
+                return (<TagBubble key={`tag-${lineageTag}`} value={lineageTag} implicit={lineageTag !== unitTag.value} level={level} />);
```
This is not rendering multiple tags properly. (BTW we should change the lightcast skills taxonomy to `allow_multiple=True` and make it apply to all orgs in the sample data.) I suggest that you make this component so that it accepts the existing object tags data for this single taxonomy as a prop, and it also has an "additional selected tags" as internal state (which will get implemented by the "Add Tags" button), then you create a function to combine these together into a tree:

```js
// appliedObjectTags is a prop, which has the data from the content_object_tags REST API, but only for this current taxonomy.
const [changedTags, setChangedTags] = React.useState(...);
const tagsTree = React.useMemo(() => {
    // compute the tree from appliedObjectTags, using the lineage
    // add in any newly added tags (that haven't yet been saved to the server) - can come in next PR
    // remove any deleted tags - can come in next PR
    // delete any "orphaned" parents with no children - can come in next PR
    return tagsTree;
}, [appliedObjectTags, changedTags]);
// render the view using tagsTree
```

**How this looks now with multiple tags:**

![Screenshot 2023-11-08 at 1 43 03 PM](https://github.com/openedx/frontend-app-course-authoring/assets/945577/8961c102-45c3-4cd9-9d76-0dd3f9aba2b3)

**How it should look:**

![Screenshot 2023-11-08 at 1 44 34 PM](https://github.com/openedx/frontend-app-course-authoring/assets/945577/7df30715-1de0-46cd-8480-f1dc0052ece8)

yusuf-musleh replied:
Oh thanks for catching this! It makes sense to create a separate component to handle this logic.

> (BTW we should change the lightcast skills taxonomy to allow_multiple=True and make it apply to all orgs in the sample data.)

Yup, I agree, I'll update the sample data script and create a PR once I've addressed all the issues.

yusuf-musleh replied:
@bradenmacdonald I've implemented this, and managed to fit it in as part of this PR https://github.com/openedx/frontend-app-course-authoring/commit/f8cbc9122e51797c2a25a0b9d6ac0f576356f446 . Your suggestions worked great, and sets up up code for the the editing PR that follows.

bradenmacdonald commented on src/unit-taxonomy-tags-drawer/TaxonomyTagsCollapsible.jsx:
```diff
@@ -0,0 +1,116 @@
+            variant="outline-primary"
+            onClick={open}
+          >
+            Add tags
```
Don't forget i18n - this should be `<FormattedMessage {...messages.addTagsButton} />`

bradenmacdonald commented on src/unit-taxonomy-tags-drawer/TaxonomyTagsCollapsible.jsx:
```diff
@@ -0,0 +1,116 @@
+  }).isRequired,
+};
+
+export default injectIntl(TaxonomyTagsCollapsible);
```
Same thing here - you're not actually using `intl` so you don't need `injectIntl`. And if you did need `intl`, it's better to use the `useIntl()` hook.

bradenmacdonald commented on src/unit-taxonomy-tags-drawer/UnitTaxonomyTagsDrawer.jsx:
```diff
@@ -0,0 +1,118 @@
+} from './api/hooks/selectors';
+import { useTaxonomyListDataResponse, useIsTaxonomyListDataLoaded } from '../taxonomy/api/hooks/selectors';
+
+const UnitTaxonomyTagsDrawer = ({ intl }) => {
```
As mentioned above, this isn't specific to units, and I'm not a big fan of "Taxonomy" in the name, so I'd recommend a simpler name like `ContentTagsDrawer`

bradenmacdonald commented on src/unit-taxonomy-tags-drawer/UnitTaxonomyTagsDrawer.jsx:
```diff
@@ -0,0 +1,118 @@
+              <hr />
+            </div>
+          ))
+          : <h3>Loading taxonomy data...</h3> }
```
Just use the `<Loading />` component to render a spinner instead of displaying this text (search the code for other examples). Otherwise, we'd have to translate this string.

yusuf-musleh replied:
I used a mix of the `<Loading />` and the `<Spinner />` components since I had issues with getting the styling to show both of them. `<Loading />` takes over the whole height of the screen.

bradenmacdonald replied:
Oh ok, sure. Whatever works! I'm actually not too familiar with these components; I just knew there should be some existing component that doesn't need text. A spinner itself is totally fine.

bradenmacdonald commented on src/unit-taxonomy-tags-drawer/TagBubble.test.jsx:
```diff
@@ -0,0 +1,86 @@
+  implicit: PropTypes.bool,
+};
+
+describe('<TagBubble />', async () => {
```
```suggestion
describe('<TagBubble />', () => {
```

This is giving a warning:

```
        Returning a Promise from "describe" is not supported. Tests must be defined synchronously.
        Returning a value from "describe" will fail the test in a future version of Jest.
```

bradenmacdonald commented on src/unit-taxonomy-tags-drawer/TaxonomyTagsCollapsible.test.jsx:
```diff
@@ -0,0 +1,83 @@
+  }).isRequired,
+};
+
+describe('<TaxonomyTagsCollapsible />', async () => {
```
```suggestion
describe('<TaxonomyTagsCollapsible />', () => {
```

bradenmacdonald commented on src/unit-taxonomy-tags-drawer/TaxonomyTagsDropDownSelector.test.jsx:
```diff
@@ -0,0 +1,102 @@
+  useTaxonomyTagsData: PropTypes.func.isRequired,
+};
+
+describe('<TaxonomyTagsDropDownSelector />', async () => {
```
```suggestion
describe('<TaxonomyTagsDropDownSelector />', () => {
```

bradenmacdonald commented on src/unit-taxonomy-tags-drawer/UnitTaxonomyTagsDrawer.test.jsx:
```diff
@@ -0,0 +1,199 @@
+  </AppProvider>
+);
+
+describe('<UnitTaxonomyTagsDrawer />', async () => {
```
```suggestion
describe('<UnitTaxonomyTagsDrawer />', () => {
```

yusuf-musleh commented on src/unit-taxonomy-tags-drawer/TagBubble.jsx:
```diff
@@ -0,0 +1,41 @@
+  level: PropTypes.number,
+};
+
+export default injectIntl(TagBubble);
```
Ah I see, thanks for the pointers. I've updated the PR with these changes.

yusuf-musleh commented:
@bradenmacdonald Thanks for the awesome review!

I've addressed most of the issues. There are 2 things remaining that I didn't quite complete today, the rendering of multiple tags properly and the renaming. 

I agree it makes sense to rename this component. I also think it's worth renaming the actual folder/files and some variable names to be more generic rather than specific to "units" and dropping the "taxonomy" wording. I think doing it now before it gets merged would save us some headache down the road.

bradenmacdonald commented:
Thanks @yusuf-musleh. I think this will be good to go once you have those done!

bradenmacdonald commented:
@yusuf-musleh BTW I really want to get this merged ASAP, so I agree that the renaming components should happen before we merge, but if the "rendering multiple tags properly" is going to take a while, let's just address that in the next PR (the editing PR or even just a simple PR focused on that one fix).

yusuf-musleh commented:
@bradenmacdonald I've opted for implementing the "rendering multiple tags properly" as a separate small PR after this one in the interest of getting this merged sooner. I've renamed the the components to be more generic, should be good to go now!

bradenmacdonald reviewed (approved):
:+1:

- [x] I tested this: in studio, as described.
- [x] I read through the code
- [x] I checked for accessibility issues
- [x] Includes documentation

yusuf-musleh commented on src/unit-taxonomy-tags-drawer/TaxonomyTagsCollapsible.jsx:
```diff
@@ -0,0 +1,116 @@
+              ariaLabel="taxonomy tags selection"
+              className="taxonomy-tags-selectable-box-set"
+            >
+              {isTaxonomyTagsLoaded && taxonomyTagsData && taxonomyTagsData.results.map((taxonomyTag) => (
```
Thought I replied to this, my bad! Yes, agreed, I've added a comment and will address it in the editing PR.

xitij2000 commented on src/content-tags-drawer/ContentTagsDropDownSelector.jsx:
```diff
@@ -0,0 +1,112 @@
+  const isOpen = (i) => (dropdownStates[i] === 1);
+
+  const clickAndEnterHandler = (index) => {
+    const updatedState = dropdownStates.map((dropdownState, i) => {
+      if (index === i) {
+        return dropdownState === 0 ? 1 : 0;
+      }
+      return dropdownState;
+    });
+
+    setDropdownStates(updatedState);
```
If I understand correctly, this is just flipping the value at index i?
Why not simplify it as follows:
```suggestion
    const updatedState = [...dropdownStates] // make a copy
    updatedState[i] ^=1; // flip 1 to 0 or 0 to 1 or use ternirary or if swithing to true false just use = !...
    setDropdownStates(updatedState);
```

xitij2000 commented on src/content-tags-drawer/ContentTagsDropDownSelector.jsx:
```diff
@@ -0,0 +1,112 @@
+  // Each index contains whether the dropdown is open/closed
+  // O === Closed
+  // 1 === Open
+  const [dropdownStates, setDropdownStates] = useState([]);
```
I think some of this code could be simplified if this an object mapping the index to true/false/undefined. 

isOpen can stay the same. 
Updating an index can be `setDropDownStates({...dropdownState, i:!dropdownState[i]})`
You don't need to initialize since initial keys are undefined which are falsy.

xitij2000 commented on src/content-tags-drawer/ContentTagsDropDownSelector.jsx:
```diff
@@ -0,0 +1,112 @@
+  return (
+    isTaxonomyTagsLoaded && taxonomyTagsData
+      ? taxonomyTagsData.results.map((taxonomyTag, i) => (
+        <div style={{ display: 'flex', flexDirection: 'column', paddingLeft: `${level * 1}rem` }}>
```
You should use classes for some of this. 
d-flex flex-column 

As for the padding, if it is nested and each keeps a margin you won't need to increase padding at each level?

yusuf-musleh replied:
I've replaced them with classes as you suggested. However I kept the `paddingLeft` since it doesn't add any margin/padding (even though it is nested) so they all appear on the same level without it.

xitij2000 commented on src/content-tags-drawer/ContentTagsDropDownSelector.jsx:
```diff
@@ -0,0 +1,112 @@
+            <SelectableBox
+              inputHidden={false}
+              type="checkbox"
+              className="pgn__selectable_box taxonomy-tags-selectable-box"
```
Why is `pgn__selectable_box` here, shouldn't it be applied automatically? Is there a bug?

yusuf-musleh replied:
You're right! I removed it. I think I initially added it there to make it more explicit since I override some of it's styles.

xitij2000 commented on src/content-tags-drawer/TagBubble.scss:
```diff
@@ -0,0 +1,9 @@
+.tag-bubble.btn-outline-dark:hover {
+  color: white;
+  background-color: #273F2F;
```
This corresponds with bg-dark, or the color $dark.

xitij2000 commented on src/content-tags-drawer/TagBubble.scss:
```diff
@@ -0,0 +1,9 @@
+}
+
+.implicit > .btn-icon-before svg {
+  color: #E9E6E4; // $light-400;
```
Why not just use $light-400?

yusuf-musleh replied:
I tried using the variables like `$light-400` but it would not work. I kept getting errors in the build that it was not defined. I'm not sure what I am doing differently compared to the other existing scss files that utilized variables?

bradenmacdonald replied:
You are importing `TagBubble.scss` into the `TagBubble` component [here](https://github.com/openedx/frontend-app-course-authoring/blob/12df9855c9b484b11ccdd72fdd61d1aa957cfd5b/src/content-tags-drawer/TagBubble.jsx#L7), which is fine, except that your `.scss` file doesn't import the SCSS variables nor Paragon base classes etc. And in fact you don't want to, because that will result in a double build and duplicated Paragon classes in the final CSS.

What you need to do instead is remove the reference to `TagBubble.scss` from your `.jsx` file, and instead include the custom styles via the master SCSS file, https://github.com/openedx/frontend-app-course-authoring/blob/master/src/index.scss . For simplicity, usually that file includes something from the top-level folder which then in turn includes the smaller .scss files.

---------

This is an "old-fashioned" system btw. The modern approach is to skip SCSS altogether and use CSS variables as well as relying much more heavily on Utility classes. I'm a big [Tailwind](https://tailwindcss.com/) fan and with that you have almost no styles at all in your global CSS and you have few or no separate CSS files. Generally you can achieve everything with utility classes like `d-flex align-self-start` etc. I don't think Paragon/boostrap's utility classes are as powerful, but the same principle applies - try to use utility classes as much as possible and create custom classes in the .css file only for very custom/unusual stuff where needed, like the Tag elements.

yusuf-musleh replied:
oh I see - I followed your suggestions and it worked! I was getting confused as to why it wasn't working. Thanks for clearing it up for me.

xitij2000 replied:
For what it's worth, once the design tokens bits land in paragon, the $light-400 will be replaced by a CSS variable anyway. There are automated tools to do that replacement in paragon alpha.

xitij2000 commented on src/content-tags-drawer/api/hooks/selectors.js:
```diff
@@ -0,0 +1,56 @@
+ */
+export const useIsTaxonomyTagsDataLoaded = (taxonomyId, fullPathProvided) => (
+  useTaxonomyTagsData(taxonomyId, fullPathProvided).status === 'success'
+);
+
+/**
+ * @returns {import("../types.mjs").ContentTaxonomyTagsData | undefined}
+ */
+export const useContentTaxonomyTagsDataResponse = (contentId) => {
+  const response = useContentTaxonomyTagsData(contentId);
+  if (response.status === 'success') {
+    return response.data;
+  }
+  return undefined;
+};
+
+/**
+ * @returns {boolean}
+ */
+export const useIsContentTaxonomyTagsDataLoaded = (contentId) => (
+  useContentTaxonomyTagsData(contentId).status === 'success'
+);
+
+/**
+ * @returns {import("../types.mjs").ContentData | undefined}
+ */
+export const useContentDataResponse = (contentId) => {
+  const response = useContentData(contentId);
+  if (response.status === 'success') {
+    return response.data.data;
+  }
+  return undefined;
+};
+
+/**
+ * @returns {boolean}
+ */
+export const useIsContentDataLoaded = (contentId) => (
+  useContentData(contentId).status === 'success'
+);
```
I'll echo the same concerns I showed [here](https://github.com/openedx/frontend-app-course-authoring/pull/645#discussion_r1377342836).

If a function starts with use is it treated as a react hook by the linters. None of these need to be react hooks, so the "use" at the beginning can be dropped.

yusuf-musleh replied:
I could be mistaken, but since the functions in `/api/hooks/api.js` return `useQuery` which is a hook, they also become hooks. If we drop the "use" in the functions in the `/api/hooks/selectors.js` that call them, I get linter errors saying that the hooks cannot be called inside normal functions and suggests to either convert them to hooks (the current case with "use") or convert them to React component functions which doesn't seem right? Maybe I am misunderstanding how to go about addressing this.

bradenmacdonald replied:
My understanding was the same as @yusuf-musleh's - any function which uses a hook is itself a hook (or a component) so should be named `use...`.

yusuf-musleh replied:
@xitij2000 @bradenmacdonald I changed the structure of these hooks/apis to follow https://github.com/openedx/frontend-app-course-authoring/pull/645/ that is using useQuery after pulling it from master. https://github.com/openedx/frontend-app-course-authoring/commit/6e1dfc72e45f2f410927363a2794c0a30c5c38e0

xitij2000 replied:
My bad, I think I got confused switching between PRs. Sorry about this, I think this looks good.

yusuf-musleh commented on src/content-tags-drawer/ContentTagsCollapsible.scss:
```diff
@@ -0,0 +1,43 @@
+  grid-auto-rows: unset !important;
+  overflow-y: scroll;
+  max-height: 20rem;
+}
```
Got it, I've updated most of them to use the paragon/bootstrap classes. https://github.com/openedx/frontend-app-course-authoring/pull/654/commits/12df9855c9b484b11ccdd72fdd61d1aa957cfd5b The width: `100%` in this case set the style for an element that is added on render from the paragon component so I could not add to the code.

yusuf-musleh commented:
@xitij2000 Thanks for the review! I addressed some of the issues, and had a few questions about some of the points.

ali-hugo commented:
Hi @yusuf-musleh,

Apologies for the late review but I only came across this PR today. I have some feedback about the styling of the tags (these changes aren't at all urgent, but it would be good if they could be implemented sometime). Please see the tables below comparing the current styling (on the left) and the correct styling (on the right):

IMPLICIT TAGS:
| Current      | Correct |
| ----------- | ----------- |
| ![implicit - current](https://github.com/openedx/frontend-app-course-authoring/assets/88187719/47000718-bf5a-4c4e-8712-90fbaa03db89) | ![implicit - correct](https://github.com/openedx/frontend-app-course-authoring/assets/88187719/8b7864dc-9101-4293-8035-f82aed856e08) |

- Tag outline should be #F2F0EF (light gray)
- Icon should be an outline and #00262B (dark green)

---

EXPLICIT TAGS:

| Current      | Correct |
| ----------- | ----------- |
| ![explicit - current](https://github.com/openedx/frontend-app-course-authoring/assets/88187719/d2825b60-3941-4022-ac4e-471fa5c60bec) | ![explicit - correct](https://github.com/openedx/frontend-app-course-authoring/assets/88187719/48a15394-2803-427c-9ee7-db0dd4b64a4a) |

- Tag outline should be #F2F0EF (light gray)

---
Please let me know if you have any questions.

yusuf-musleh commented:
@ali-hugo Thanks for the review and feedback. I made the changes regarding the Tag outline color.

However for the implicit icon, the outline tag icon is not available in [Paragon](https://paragon-openedx.netlify.app/foundations/icons/) or in the [Material Symbols and Icons from Google Fonts](https://fonts.google.com/icons?icon.query=tag) (that is used under the hood in Paragon). That's why i went with the different color that didn't match the design.

However I did [find it](https://fonts.google.com/icons?icon.query=tag) as part of a newer version of font-awesome (6.4.2) but it is considered a "Pro" icon. I'm not sure what is the general process is in this situation when the icon is not available? Have you faced this before @bradenmacdonald ?

ali-hugo commented:
@yusuf-musleh Thanks for updating the outline color 🙂 .

The outline tag icon _is_ actually available in the Google Fonts icons - it's just a bit tricky to find. You have to ensure that the "fill" toggle in the filters drawer is off. You should be able to access the outline version of the icon at [this link](https://fonts.google.com/icons?selected=Material%20Symbols%20Outlined%3Asell%3AFILL%400%3Bwght%40400%3BGRAD%400%3Bopsz%4024). I have also included the svg below:

![sell_FILL0_wght400_GRAD0_opsz24](https://github.com/openedx/frontend-app-course-authoring/assets/88187719/b3a06420-60ab-49bf-ac27-0bc17fb6b7e1)

Let me know if you run into any trouble.

yusuf-musleh commented:
@ali-hugo ohh! I completely missed that! Thanks for pointing it out. I've added it as a custom tag and updated the style, here's how it looks like now (the filled in one is hovered on):

![Screen Shot 2023-11-16 at 3 14 43 PM](https://github.com/openedx/frontend-app-course-authoring/assets/6829768/88898895-21d8-46b9-9a29-479423d93f8e)

Ideally we'd create a PR and add the icon to Paragon, but we can leave that for a later task/ticket.

ali-hugo commented:
@yusuf-musleh Yay, that looks great! Thank you 😄 

> Ideally we'd create a PR and add the icon to Paragon, but we can leave that for a later task/ticket.

Perhaps we should wait until we've done some user testing before we add the icon to Paragon. That way we can make sure that the chosen icon makes sense to users before we make it a permanent feature. Users never fail to surprise me with their feedback!

xitij2000 commented on src/content-tags-drawer/ContentTagsCollapsible.jsx:
```diff
@@ -0,0 +1,115 @@
+              type="checkbox"
+              name="tags"
+              columns={1}
+              ariaLabel="taxonomy tags selection"
```
This should be internationalised.

xitij2000 commented on src/content-tags-drawer/ContentTagsCollapsible.scss:
```diff
@@ -0,0 +1,32 @@
+}
+
+.tags-count-badge {
+  margin-top: 1rem;
```
This is equivalent to `mt-3`

yusuf-musleh replied:
Updated https://github.com/openedx/frontend-app-course-authoring/pull/654/commits/1c9d0f8c03cc3ab8a41a192426c8bc43792705c1

xitij2000 commented on src/content-tags-drawer/ContentTagsCollapsible.scss:
```diff
@@ -0,0 +1,32 @@
+}
+
+.tags-count-badge-empty {
+  visibility: hidden;
```
= `.invisble`

yusuf-musleh replied:
Updated https://github.com/openedx/frontend-app-course-authoring/pull/654/commits/1c9d0f8c03cc3ab8a41a192426c8bc43792705c1

xitij2000 commented on src/content-tags-drawer/ContentTagsDrawer.jsx:
```diff
@@ -0,0 +1,128 @@
+
+  const closeContentTagsDrawer = () => {
+    // "*" allows communication with any origin
+    window.parent.postMessage('closeManageTagsDrawer', '*');
```
I don't there is any set standard for this, but a few other places this is used it's a string like: 'lmshtmlfragment.resize' and 'tools.toggleNotes'.

yusuf-musleh replied:
Make this change would require also making changes to the edx-platform. I think so it doesn't get blocked by another review cycle there, we can address it in a later PR.

xitij2000 commented on src/content-tags-drawer/ContentTagsDrawer.jsx:
```diff
@@ -0,0 +1,128 @@
+    const handleEsc = (event) => {
+      /* Close drawer when ESC-key is pressed and selectable dropdown box not open */
+      const targetElement = event.target;
+      const selectableBox = targetElement.classList.contains('pgn__selectable_box');
```
Could there be a better way of doing this? For instance perhaps adding a data attribute and matching that?

yusuf-musleh replied:
While addressing this, I found a bug with closing the drawer when the dropdown is open. I updated it with a data-attr and fixed the issue. https://github.com/openedx/frontend-app-course-authoring/pull/654/commits/cda3a9179ecf2211c9bddafd71c5d5d402bb9504

xitij2000 commented on src/content-tags-drawer/ContentTagsDropDownSelector.jsx:
```diff
@@ -0,0 +1,99 @@
+    setDropdownStates({ ...dropdownStates, [i]: !dropdownStates[i] });
+  };
+
+  const useTaxonomyTagsData = (id, fullPathProvided) => {
+    const taxonomyTagsData = useTaxonomyTagsDataResponse(id, fullPathProvided);
+    const isTaxonomyTagsLoaded = useIsTaxonomyTagsDataLoaded(id, fullPathProvided);
+    return { taxonomyTagsData, isTaxonomyTagsLoaded };
+  };
+
+  const { taxonomyTagsData, isTaxonomyTagsLoaded } = useTaxonomyTagsData(taxonomyId, subTagsUrl);
```
Won't it be shorter to remove the function call here?

```suggestion
  const taxonomyTagsData = useTaxonomyTagsDataResponse(taxonomyId, subTagsUrl);
  const isTaxonomyTagsLoaded = useIsTaxonomyTagsDataLoaded(taxonomyId, subTagsUrl);
```

yusuf-musleh replied:
Yup! Updated it here: https://github.com/openedx/frontend-app-course-authoring/pull/654/commits/43b2de57da7bfcd8b3442d7b41cb961c76ab79ff

xitij2000 commented on src/content-tags-drawer/TagOutlineIcon.jsx:
```diff
@@ -0,0 +1,21 @@
+const TagOutlineIcon = (props) => (
+  <svg
+    xmlns="http://www.w3.org/2000/svg"
+    height="24px"
+    viewBox="0 0 24 24"
+    width="24px"
+    fill="currentColor"
+    role="img"
+    focusable="false"
+    aria-hidden="true"
+    {...props}
+  >
+    <path d="M0 0h24v24H0V0z" fill="none" />
+    <path
+      d="m21.41 11.58-9-9C12.05 2.22 11.55 2 11 2H4c-1.1 0-2 .9-2 2v7c0 .55.22 1.05.59 1.42l9 9c.36.36.86.58 1.41.58s1.05-.22 1.41-.59l7-7c.37-.36.59-.86.59-1.41s-.23-1.06-.59-1.42zM13 20.01 4 11V4h7v-.01l9 9-7 7.02z"
+    />
+    <circle cx="6.5" cy="6.5" r="1.5" />
+  </svg>
+);
+
+export default TagOutlineIcon;
```
This should be moved to paragon right?

yusuf-musleh replied:
Ideally yes, since this icon is not already in Paragon. But we've agreed to create a PR in paragon to add it once we've confirmed with the user that they liked it. https://github.com/openedx/frontend-app-course-authoring/pull/654#issuecomment-1814350972

xitij2000 commented on src/taxonomy/data/api.js:
```diff
@@ -3,18 +3,27 @@ import { camelCaseObject, getConfig } from '@edx/frontend-platform';
 const getApiBaseUrl = () => getConfig().STUDIO_BASE_URL;
-export const getTaxonomyListApiUrl = () => new URL('api/content_tagging/v1/taxonomies/?enabled=true', getApiBaseUrl()).href;
+
+export const getTaxonomyListApiUrl = (org) => {
+  let queryParams = '?enabled=true';
+  if (org !== undefined) {
+    queryParams = `${queryParams}&org=${org}`;
+  }
+  return new URL(`api/content_tagging/v1/taxonomies/${queryParams}`, getApiBaseUrl()).href;
+};
+
```
```suggestion
export const getTaxonomyListApiUrl = (org) => {
  const url = new URL(`api/content_tagging/v1/taxonomies/${queryParams}`, getApiBaseUrl());
  url.searchParams.append('enabled', true);
  if (org !== undefined) {
    url.searchParams.append('org', org);
  }
  return url.href;
};

```

yusuf-musleh replied:
Much cleaner! Updated: https://github.com/openedx/frontend-app-course-authoring/pull/654/commits/dd9ef5a38f5f46bd8bb837f7468c01d7342ffe71

xitij2000 reviewed (approved):
I have mostly just minor nits, other than that it looks good to go.

yusuf-musleh commented:
@xitij2000 Thanks for your review and great recommendations! I've addressed them and updated the PR.

@bradenmacdonald This should be good go!

bradenmacdonald commented:
@xitij2000 Thanks for the review! @yusuf-musleh I can't merge this; we need @xitij2000  to hit merge once he's happy with it. Though at this point I'd appreciate pushing any minor issues to future PRs, as we need to get this merged to unblock other work. We let this PR get too big.

xitij2000 commented:
@yusuf-musleh Could you squash this down to one or a few commits with a descriptive commit message? I will merge ASAP after that.

yusuf-musleh commented:
@xitij2000 I squashed the commits and went through it another time locally, should be good to go to merge.

xitij2000 commented:
> @xitij2000 I squashed the commits and went through it another time locally, should be good to go to merge.

Could you add a description of the commit that describes what it's adding?

yusuf-musleh commented:
> > @xitij2000 I squashed the commits and went through it another time locally, should be good to go to merge.
> 
> Could you add a description of the commit that describes what it's adding?

Sure thing, updated.

openedx-webhooks commented:
@yusuf-musleh 🎉 Your pull request was merged!  Please take a moment  to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=openedx/frontend-app-course-authoring&entry.867055334=https://github.com/openedx/frontend-app-course-authoring/pull/654&entry.1484655318=https://github.com/yusuf-musleh&entry.752974735=2023-10-23+14:21&entry.1917517419=2023-11-20+19:56&entry.2133058324=Yes) so we can improve your experience in the future.

<!-- comment:end_survey -->
